segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/hdfs/HDFSStorage.java

```diff
@@ -196,86 +328,302 @@ public boolean supportsTruncation() {
+                // Update our knowledge of the filesystem and throw a BadOffsetException - this should cause upstream code
+                // to try to reconcile; if it can't then the upstream code should shut down or take other appropriate measures.
+                log.warn("File changed detected for '{}'. Expected length = {}, actual length = {}.", lastFile, lastFile.getLen(), stream.getPos());
+                throw new BadOffsetException(handle.getSegmentName(), lastFile.getLen(), offset);
+            }
+
+            if (length == 0) {
+                // Exit here (vs at the beginning of the method), since we want to throw appropriate exceptions in case
+                // of Sealed or BadOffset
+                // Note: IOUtils.copyBytes with length == 0 will enter an infinite loop, hence the need for this check.
+                return;
+            }
+
+            // We need to be very careful with IOUtils.copyBytes. There are many overloads with very similar signatures.
+            // There is a difference between (InputStream, OutputStream, int, boolean) and (InputStream, OutputStream, long, boolean),
+            // in that the one with "int" uses the third arg as a buffer size, and the one with "long" uses it as the number
+            // of bytes to copy.
+            IOUtils.copyBytes(data, stream, (long) length, false);
+
+            stream.flush();
+        } catch (IOException ex) {
+            throw HDFSExceptionHelpers.throwException(handle.getSegmentName(), ex);
+        }
+
+        HDFSMetrics.WRITE_LATENCY.reportSuccessEvent(timer.getElapsed());
+        HDFSMetrics.WRITE_BYTES.add(length);
+        LoggerHelpers.traceLeave(log, "write", traceId, handle, offset, length);
     }
 
     @Override
-    public SegmentProperties getStreamSegmentInfo(String streamSegmentName) throws StreamSegmentException {
-        return call(new GetInfoOperation(streamSegmentName, this.context));
+    public SegmentHandle openWrite(String streamSegmentName) throws StreamSegmentException {
+        long traceId = LoggerHelpers.traceEnter(log, "openWrite", streamSegmentName);
+        int fencedCount = 0;
+        do {
+            try {
+                FileStatus existingFiles = findStatusForSegment(streamSegmentName, true);
+
+                if (!isSealed(existingFiles.getPath())) {
+                    if (getEpocFromPath(existingFiles.getPath()) > this.epoch) {
+                        throw new StorageNotPrimaryException(streamSegmentName);
+                    }
+
+                    Path targetPath = getFilePath(streamSegmentName, this.epoch);
+                    if (!targetPath.equals(existingFiles.getPath())) {
+                        try {
+                            this.fileSystem.rename(existingFiles.getPath(), targetPath);
+                        } catch (FileNotFoundException e) {
+                            //This happens when more than one host is trying to fence and only one of the host goes through
```

Comment:
There are two sentences in here. Please use punctuation and capitalization to reflect that.